File: go/pkg/service/task_service.go
package service

import (
	"errors"
	"fmt"
	"github.com/taskmanager/go-client/pkg/model"
	"github.com/taskmanager/go-client/pkg/util"
)

// TaskService handles business logic for tasks
// Dependencies:
//   - model package (Task, Priority)
//   - util package (NativeUtils - references native/utils.h)
type TaskService struct {
	tasks    []*model.Task
	utils    *util.NativeUtils
	validator *util.TaskValidator
}

// NewTaskService creates a new TaskService instance
func NewTaskService() *TaskService {
	return &TaskService{
		tasks:     make([]*model.Task, 0),
		utils:     util.NewNativeUtils(),
		validator: util.NewTaskValidator(),
	}
}

// CreateTask creates a new task with validation
func (s *TaskService) CreateTask(title, description string, priority model.Priority) (*model.Task, error) {
	if !s.validator.ValidateTitle(title) {
		return nil, errors.New("invalid task title")
	}

	newID := s.getNextID()
	
	// Use native utilities to calculate priority score
	priorityScore := s.utils.CalculatePriorityScore(title, description)
	adjustedPriority := s.adjustPriorityByScore(priority, priorityScore)

	task := &model.Task{
		ID:          newID,
		Title:       title,
		Description: description,
		Completed:   false,
		Priority:    adjustedPriority,
	}

	s.tasks = append(s.tasks, task)
	return task, nil
}

// GetAllTasks returns all tasks
func (s *TaskService) GetAllTasks() []*model.Task {
	return s.tasks
}

// GetTaskByID returns a task by its ID
func (s *TaskService) GetTaskByID(id int) (*model.Task, error) {
	for _, task := range s.tasks {
		if task.ID == id {
			return task, nil
		}
	}
	return nil, fmt.Errorf("task not found: %d", id)
}

// UpdateTask updates an existing task
func (s *TaskService) UpdateTask(id int, update *model.TaskUpdate) (*model.Task, error) {
	task, err := s.GetTaskByID(id)
	if err != nil {
		return nil, err
	}

	if update.Title != nil {
		if !s.validator.ValidateTitle(*update.Title) {
			return nil, errors.New("invalid task title")
		}
		task.Title = *update.Title
	}

	if update.Description != nil {
		task.Description = *update.Description
	}

	if update.Completed != nil {
		task.Completed = *update.Completed
	}

	if update.Priority != nil {
		task.Priority = *update.Priority
	}

	return task, nil
}

// DeleteTask deletes a task by ID
func (s *TaskService) DeleteTask(id int) error {
	for i, task := range s.tasks {
		if task.ID == id {
			s.tasks = append(s.tasks[:i], s.tasks[i+1:]...)
			return nil
		}
	}
	return fmt.Errorf("task not found: %d", id)
}

// GetTasksByPriority returns all tasks with a specific priority
func (s *TaskService) GetTasksByPriority(priority model.Priority) []*model.Task {
	var result []*model.Task
	for _, task := range s.tasks {
		if task.Priority == priority {
			result = append(result, task)
		}
	}
	return result
}

// GetStatistics returns task statistics
func (s *TaskService) GetStatistics() *TaskStatistics {
	total := len(s.tasks)
	completed := 0
	for _, task := range s.tasks {
		if task.Completed {
			completed++
		}
	}

	// Use native factorial utility
	factorialTotal := s.utils.Factorial(total)

	return &TaskStatistics{
		TotalTasks:    total,
		CompletedTasks: completed,
		PendingTasks:   total - completed,
		FactorialTotal: factorialTotal,
	}
}

func (s *TaskService) getNextID() int {
	maxID := 0
	for _, task := range s.tasks {
		if task.ID > maxID {
			maxID = task.ID
		}
	}
	return maxID + 1
}

func (s *TaskService) adjustPriorityByScore(original model.Priority, score int) model.Priority {
	if score > 50 && original != model.PriorityHigh {
		return model.PriorityHigh
	} else if score < 20 && original != model.PriorityLow {
		return model.PriorityLow
	}
	return original
}

File: java/src/main/java/com/taskmanager/service/TaskService.java
package com.taskmanager.service;

import com.taskmanager.model.Task;
import com.taskmanager.model.Priority;
import com.taskmanager.util.NativeUtils;
import com.taskmanager.util.TaskValidator;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * TaskService - Business logic for task management
 * Dependencies:
 *   - Task (model)
 *   - Priority (model)
 *   - NativeUtils (utility - references native/utils.h)
 *   - TaskValidator (utility)
 */
public class TaskService {
    private List<Task> tasks;
    private TaskValidator validator;
    private NativeUtils nativeUtils;

    public TaskService() {
        this.tasks = new ArrayList<>();
        this.validator = new TaskValidator();
        this.nativeUtils = new NativeUtils();
    }

    /**
     * Create a new task with validation and priority scoring
     */
    public Task createTask(String title, String description, Priority priority) {
        if (!validator.validateTitle(title)) {
            throw new IllegalArgumentException("Invalid task title");
        }

        int newId = tasks.isEmpty() ? 1 : tasks.stream()
                .mapToInt(Task::getId)
                .max()
                .orElse(0) + 1;

        Task task = new Task(newId, title, description, false, priority);
        
        // Use native utility to calculate priority score
        int priorityScore = nativeUtils.calculatePriorityScore(title, description);
        task.setPriority(adjustPriorityByScore(priority, priorityScore));
        
        tasks.add(task);
        return task;
    }

    /**
     * Get all tasks
     */
    public List<Task> getAllTasks() {
        return new ArrayList<>(tasks);
    }

    /**
     * Get task by ID
     */
    public Task getTaskById(int id) {
        return tasks.stream()
                .filter(t -> t.getId() == id)
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Task not found: " + id));
    }

    /**
     * Update task
     */
    public Task updateTask(int id, String title, String description, Boolean completed, Priority priority) {
        Task task = getTaskById(id);
        
        if (title != null) {
            if (!validator.validateTitle(title)) {
                throw new IllegalArgumentException("Invalid task title");
            }
            task.setTitle(title);
        }
        
        if (description != null) {
            task.setDescription(description);
        }
        
        if (completed != null) {
            task.setCompleted(completed);
        }
        
        if (priority != null) {
            task.setPriority(priority);
        }
        
        return task;
    }

    /**
     * Delete task
     */
    public void deleteTask(int id) {
        Task task = getTaskById(id);
        tasks.remove(task);
    }

    /**
     * Get tasks by priority
     */
    public List<Task> getTasksByPriority(Priority priority) {
        return tasks.stream()
                .filter(t -> t.getPriority() == priority)
                .collect(Collectors.toList());
    }

    /**
     * Get completed tasks
     */
    public List<Task> getCompletedTasks() {
        return tasks.stream()
                .filter(Task::isCompleted)
                .collect(Collectors.toList());
    }

    /**
     * Calculate statistics using native utilities
     */
    public TaskStatistics getStatistics() {
        int total = tasks.size();
        int completed = (int) tasks.stream().filter(Task::isCompleted).count();
        
        // Use native factorial for some calculation
        long factorialTotal = nativeUtils.factorial(total);
        
        return new TaskStatistics(total, completed, total - completed, factorialTotal);
    }

    private Priority adjustPriorityByScore(Priority original, int score) {
        if (score > 50 && original != Priority.HIGH) {
            return Priority.HIGH;
        } else if (score < 20 && original != Priority.LOW) {
            return Priority.LOW;
        }
        return original;
    }
}

